Restore pycompat API

lehmannro opened the pull request:
As discussed in #1726, restore all helper functions in `sphinx.util.pycompat`.  I have also added a deprecation warning to the functions.

lehmannro committed:
Restore pycompat API. Fixes #1726.

All compatibility helpers which have been in this module should be
resurrected -- and, at the same time, deprecated -- to grant
backwards-compatibility with older Sphinx versions to third-party
extensions.

This does NOT restore compatibility with any older Python versions.

shimizukawa commented on sphinx/util/pycompat.py:
```diff
@@ -103,3 +103,33 @@ def execfile_(filepath, _globals):
+    def wrapped(*args, **kwargs):
+        warnings.warn("sphinx.util.pycompat.{0.__name__} is deprecated, "
+                      "please use the standard library version.".format(func),
+                      DeprecationWarning, stacklevel=2)
```
It would be better if we indicate removing version for deprecated names.

lehmannro replied:
Which version do you suggest?  1.4?

shimizukawa replied:
I suggest 1.4.

lehmannro replied:
Added to the warning text.  I briefly thought about making the version configurable on a per-deprecation basis, but since we only have this one set of deprecations right now…

shimizukawa replied:
Agreed.

lehmannro committed:
Warn when accessing deprecated attributes.

Previously, only calling a deprecated function issued a warning.
Attributes (like sphinx.util.pycompat.class_types) would never warn.

lehmannro committed:
Also restore pycompat.{String,Bytes}IO.

lehmannro committed:
Slated for removal in 1.4.

shimizukawa commented on sphinx/util/pycompat.py:
```diff
@@ -103,3 +103,40 @@ def execfile_(filepath, _globals):
+
+import warnings
+
+from six import class_types
```
I guess class_types is not used.

lehmannro replied:
It seems some people used to use it. [[1]](https://bitbucket.org/bwanamarko/sphinxcontrib-matlabdomain/pull-request/1/enh-remove-unused-import-of-base_exception/diff) [[2]](https://github.com/CybOXProject/python-cybox/blob/645c419/cybox/utils/autoentity.py#L3) [[3]](https://github.com/earthreader/libearth/blob/bc9a25898e/docs/conf.py#L39)

If we conclude that's never been part of the public API we should also remove `base_exception` I guess.

shimizukawa replied:
As long as it's used, it's also marked as deprecation I think.

shimizukawa replied:
@lehmannro How do you think for my comment? If you are ok, I'm going to merge this for 1.3b3.

lehmannro replied:
I am not sure — do you have any change proposals?  Otherwise I'll just merge it into master tomorrow (+ a changelog entry.)

shimizukawa replied:
I have one proposal. 
`class_types` should be also deprecated. It seems that there is no reason to exclude `class_types` from deprecation.

(I think it is not a big issue, it just a meaning of my first comment. sorry for less explanation.)

birkenfeld commented:
If this works for the use case, feel free to merge it.

shimizukawa commented:
LGTM with nit. @lehmannro can you please proceed to merge?

lehmannro committed:
Also restore pycompat.class_types and deprecate it properly.

lehmannro committed:
Mention pycompat in changelog.

shimizukawa commented:
Thanks!
